### PR TITLE
Added subdomain dns check for domain verification.

### PR DESCRIPTION
**Pull request by Randy808**
Currently domains are validated using a well-known uri located at `https://{domain}/.well-known/liquid-asset-issuer-proof-{asset_id}` but this change aims to add an alternative verification method leveraging DNS TXT records. This would empower issuers to add proofs using their DNS provider and avoid the requirement of hosting at a particular domain.

### CODE REVIEW

**Review by shesek — CHANGES_REQUESTED** [review]

**Comment by shesek** [comment]
Why was it decided to verify the `TXT` record against a subdomain with the asset's id prefix?

Requiring issuers to prove their ability to add `TXT` records directly to the user-visible domain name would give us a stronger guarantee of ownership. The distinction shouldn't matter in the typical case, but there are some exceptional cases where it may - for example domain owners that give DNS control over subdomains to their users (like afaird.org or dyno), or if an attacker manages to gain DNS access to a seemingly-innocent subdomain through social engineering (another way to mitigate the last concern is to make the subdomain name intention clearer, for example `liquid-asset-verification-d8a5e71fab92.foocorp.com`).

Note that verifying against the main domain name requires checking all `TXT` records and not just the first one as currently implemented, both because they may be other non-liquid-related `TXT` records and because there may be multiple assets linked to the domain.

This also has the advantage of discoverability, as users can query for `TXT` records from the main domain name to find linked assets, without having to know the asset ids to begin with. This can also be used to resolve strings like `usdt@tether.to` to the linked asset id (tickers are guaranteed to be unique within a domain name). Doing both things is already theoretically possible through the asset registry (if it supported lookups by domain name, which it currently doesn't), but it can reduce the reliance on it.

**Comment by shesek** [comment]
It would be awesome if this used @TheBlueMatt's [`dnssec-prover`](https://github.com/TheBlueMatt/dnssec-prover) ([website](https://http-dns-prover.as397444.net/)) to keep and serve self-validating proofs of domain ownership/linkage.

With the current HTTP-based verification method, we basically tell users that we looked and found the verification file at some point in the past, but can't prove it to them. The user has to take our word for it, or issue an HTTP request to check themselves - which requires costly IO (esp if you're looking to verify the registry in bulk) and may fail due to some temporary error, or because the file was since removed. If it does fail, the user can't tell if we lied or if something changed since. And even if they can somehow tell, they wouldn't be able to prove.

With `dnssec-prover` we can provide proofs that don't require trusting the asset registry (only the DNS root trust anchors), and that can be validated on their own without requiring any network activity (offline even).

**Comment by Randy808** [comment]
I liked the fixed subdomain approach. I considered putting the TXT records on the main domain but I didn't want to interfere with any existing TXT records infrastructure an issuer may have already set up. I also think discoverability of assets for a particular domain is desirable over downloading and searching the index.json file from [asset_registry_db](https://github.com/Blockstream/asset_registry_db).

I'll also look more into the [dnssec-prover](https://github.com/TheBlueMatt/dnssec-prover) you mentioned because it sounds like there would be a lot of upside for issuers if implemented.

**Comment by shesek** [comment]
Thanks for better explaining the motivation. So basically that issuers don't have to manage a webserver in addition to the DNS they must already manage either way. Makes sense.

---

I suggest we use a a more compact verification string for the TXT record rather than the human-readable sentence we use for HTTP-based verification. And also include the ticker name in there. Perhaps something like `liquid-asset-verification=<asset-id>,<ticker>`

---

It should be noted that the TXT record only verifies the linkage in one direction - that the domain owner verifiably agreed to associate their domain name to the specified asset id (and ticker if we add that). But it doesn't verify that the on-chain issuance associated itself with the domain name via the contract hash commitment.

I think there's value in providing just that one-way verification (typically the trust anchor for the user is the domain name and not the asset id, so while it makes sense for an asset to try and impersonate itself as related to a domain, it doesn't really seem to make sense for a domain owner to impersonate an asset id), but we could theoretically do better if we wanted to -

A two-way verification requires 3 more pieces of information: the `contract_json` (and its SHA256 `contract_hash`), the `issuance_txin` (asset issuance txid + the index of the issuance input in it) and the `issuance_prevout` (the prev output spent by the issuance input)

The `asset_id` is a function of `contract_hash`, `issuance_txin`  and `issuance_prevout`, so given these 3 you can verify that the `asset_id` issuance committed to the given contract json/hash.

We can get the `asset_id`, `contract_hash`, `issuance_txin` and `issuance_prevout` to fit inside a TXT record (its ASCII-only with a maximum of 255 characters, and we need roughly 183 characters if we use bae64 encoding. also the `asset_id` isn't strictly necessary because it can be computed from the other fields, but its probably better to have it more readily available. and you don't really need the `contract_hash` if you have the full `contract_json`, however...)

However getting the `contract_json` in is more tricky. It doesn't quite fit even as a separate TXT record. I checked and (after disregarding some extreme outliers in large volumes that were nearly identical and significantly altered the results) it appears that 68% of the issuance JSONs are under the 255 limit while the average issuance JSON length is 281.

A simple solution is splitting the `contract_json` over multiple TXT records. Another one is restructuring the contract json hashing such that we only hash only the officially supported fields at the top level (version, ticker, name, precision, domain name and issuer pubkey), plus a single hash for all the custom fields added by the issuer if any (could also be a merkle tree but that may be an overkill). That way the proof data necessary for verifying the official fields (and specifically the domain linkage) is more compact and should fit inside a single record. You'll only need the full `contract_json` if you need to verify custom fields.

**Comment by shesek** [comment]
> I think the two-way verification is interesting but I don't want it to delay this PR.

Agreed. Just wanted to share my thoughts and put this out there as an option, I wouldn't prioritize this either. (It's perhaps worth noting that other similar domain-based solutions like Lightning Address and TheBlueMatt's [Human Readable Payment Instructions](https://delvingbitcoin.org/t/human-readable-bitcoin-payment-instructions/542) don't do two-way either)

So we just need to finalize the DNS address to use and the TXT record contents. Also `dnssec-prover` would be a big win imo, but it can be done later in a separate PR.

**Comment by Randy808** [comment]
Update the PR but haven't been able to get the tests to run

**Comment by Randy808** [comment]
@shesek Anything else needed to get this PR in?

**Review by shesek — CHANGES_REQUESTED** [review]
I'm not so sure about @TheBlueMatt's suggestion (I think it might've been tongue in cheek? like *if you're not doing it the right way, you might as well..*)

For me, using the standard DNS protocol is preferable over using Google's `dns.google` even if we're using Google's `8.8.8.8`, as there's no vendor lock-in and we could freely and easily switch to another DNS server at any time.

If using Google's DNS servers isn't desirable, we could also use [`Resolver::from_system_conf()`](https://docs.rs/trust-dns-resolver/latest/trust_dns_resolver/struct.Resolver.html#method.from_system_conf), which will use the DNS servers from `/etc/resolve.conf` (probably the datacenter's DNS?), or define any other DNS server we want.

Also, if we do end up deciding to use DNS, perhaps we could look into switching from `trust_dns_resolver` to `hickory_resolver`? The former mentions on its README that the latter is its continuation.

**Comment by shesek** [comment]
> If using Google's DNS servers isn't desirable, we could also use `Resolver::from_system_conf()`, which will use the DNS servers from `/etc/resolve.conf`

Oh huh, I just realized that this would actually also be Google, since this is hosted on GCP...

File: src/entity.rs [hunk 1]
```diff
@@ -90,39 +88,80 @@ fn verify_domain_link_http(asset: &Asset, domain: &str) -> Result<()> {
...
+        asset_id,
+        asset.fields.ticker.clone().unwrap_or(String::from(""))
     );
 
+    let labels = domain.split('.').collect::<Vec<&str>>();
+    ensure!(labels.len() > 1, "domain must have at least two labels");
+
+    let root_domain = labels[labels.len() - 2];
+    let tld = labels[labels.len() - 1];
+    let root_domain = format!("{}.{}", root_domain, tld);
```
[the remark this answers is not in the document]
Comment (Randy808) [reply src/entity.rs:139]: Wasn't aware that the UK [used to restrict registration under a subdomain](https://serverfault.com/a/595424/1052341). Normally I'd imagine an owner of a subdomain to have access to the root unless the root was providing a service to add CNAMEs to subdomains ([which disqualifies having TXT records](https://stackoverflow.com/a/34614769/7407434)), but .co.uk is an example to the counter.

**Comment by Randy808** [comment]
> I'm not so sure about @TheBlueMatt's suggestion (I think it might've been tongue in cheek? like _if you're not doing it the right way, you might as well.._)
> 
> For me, using the standard DNS protocol is preferable over using Google's `dns.google` even if we're using Google's `8.8.8.8`, as there's no vendor lock-in and we could freely and easily switch to another DNS server at any time.
> 
> If using Google's DNS servers isn't desirable, we could also use [`Resolver::from_system_conf()`](https://docs.rs/trust-dns-resolver/latest/trust_dns_resolver/struct.Resolver.html#method.from_system_conf), which will use the DNS servers from `/etc/resolve.conf` (probably the datacenter's DNS?), or define any other DNS server we want.
> 
> Also, if we do end up deciding to use DNS, perhaps we could look into switching from `trust_dns_resolver` to `hickory_resolver`? The former mentions on its README that the latter is its continuation.

No I don't think he was considering [the default that's used for hickory-dns (and trust-dns) is google ](https://github.com/hickory-dns/hickory-dns/blob/cffc3fac2a7dc75e6a56b2a348033b5be903fdde/crates/resolver/src/config.rs#L309). The newer versions of hickory-dns had some kind of dependency conflict so trust-dns was used to avoid having to making any changes to dependency versions (which are already extremely out of date given that there hasn't been a commit in 2 years).

File: src/entity.rs [hunk 2]
```diff
@@ -90,39 +88,80 @@ fn verify_domain_link_http(asset: &Asset, domain: &str) -> Result<()> {
...
+    let txt_records = txt_lookup(root_domain)?;
 
-    match txt_records.iter().any(|record| {
-        let raw_txt_data = record.txt_data();
-        match str::from_utf8(&raw_txt_data[0]) {
-            Ok(parsed_body) => parsed_body.trim_end() == expected_body,
-            Err(_) => false
-        }
-    }) {
+    match txt_records
```
[the remark this answers is not in the document]
Comment (Randy808) [reply src/entity.rs]: Changed to `if` statement

File: src/entity.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Randy808) [reply src/entity.rs]: Changed to use full domain

File: src/entity.rs [hunk 3]
```diff
@@ -90,39 +88,80 @@ fn verify_domain_link_http(asset: &Asset, domain: &str) -> Result<()> {
...
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct TxtRecord {
+    name: String,
+    #[serde(rename = "type")]
+    dns_type: u32,
+    #[serde(rename = "TTL")]
+    ttl: u32,
+    data: String
```
[the remark this answers is not in the document]
Comment (Randy808) [reply src/entity.rs]: Removed struct

**Comment by Randy808** [comment]
Updated PR 
cc: @jtrivinop

**Comment by jtrivinop** [comment]
@shesek is there anything pending here before this can be merged?

**Review by philippem — APPROVED** [review]

**Comment by shesek** [comment]
I'm still not convinced that avoiding an extra dependency is worth relying on a non-standard Google-only API . (we have another PR that adds 6 new ones...)

If it is decided to go with this approach, the implementation itself looks good to me. But there appears to be some unintentionally committed changes to `Cargo.toml` and the `Dockerfile`.

**Comment by Randy808** [comment]
Not mistaken, but reverted since #52 seems to do a more comprehensive update.

**Review by philippem — APPROVED** [review]

**Pull request merged**